File: Examples2018/UIAlertDialog/app/src/main/java/course/examples/ui/alertdialog/AlertDialogActivity.java
package course.examples.ui.alertdialog;

import android.app.Activity;
import android.app.DialogFragment;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;

public class AlertDialogActivity extends Activity {

    // Identifier for each type of Dialog
    private static final int ALERTTAG = 0, PROGRESSTAG = 1;

    private static final String TAG = "AlertDialogActivity";
    private Button mShutdownButton = null;
    private DialogFragment mDialog;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);

        // ShutDown Button
        mShutdownButton = findViewById(R.id.shutdownButton);
    }

    public void onShutDownButtonClicked(@SuppressWarnings("unused") View view) {
        showDialogFragment(ALERTTAG);
    }

    // Show desired Dialog
    private void showDialogFragment(int dialogID) {

        switch (dialogID) {

            // Show AlertDialog
            case ALERTTAG:

                // Create a new AlertDialogFragment
                mDialog = AlertDialogFragment.newInstance();

                // Show AlertDialogFragment
                mDialog.show(getFragmentManager(), "Alert");

                break;

            // Show ProgressDialog
            case PROGRESSTAG:

                // Create a new ProgressDialogFragment
                mDialog = ProgressDialogFragment.newInstance();

                // Show new ProgressDialogFragment
                mDialog.show(getFragmentManager(), "Shutdown");
                break;
        }
    }

    // Abort or complete ShutDown based on value of shouldContinue
    void continueShutdown(boolean shouldContinue) {
        if (shouldContinue) {

            // Prevent further interaction with the ShutDown Button
            mShutdownButton.setEnabled(false);

            // Show ProgressDialog as shutdown process begins
            showDialogFragment(PROGRESSTAG);

            // Finish the ShutDown process
            finishShutdown();

        } else {

            // Abort ShutDown and dismiss dialog
            mDialog.dismiss();
        }
    }

    private void finishShutdown() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    // Pretend to do something before
                    // shutting down
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    Log.i(TAG, e.toString());
                } finally {
                    finish();
                }
            }
        }).start();
    }

}

File: ExamplesKotlin/UIAlertDialog/app/src/main/java/course/examples/ui/alertdialog/AlertDialogActivity.kt
package course.examples.ui.alertdialog

import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentActivity
import android.view.View
import android.widget.Button
import android.widget.ProgressBar
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class AlertDialogActivity : FragmentActivity() {

    companion object {

        // Identifier for each type of Dialog
        private const val ALERT_ID = 0
        private const val PROGRESS_ID = 1

        @Suppress("unused")
        private const val TAG = "AlertDialogActivity"
        private const val ALERT_TAG = "AlertDialog"
    }

    private lateinit var mShutdownButton: Button
    private lateinit var mDialog: DialogFragment
    private lateinit var mProgressButton: ProgressBar

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.main)

        // ShutDown Button
        mShutdownButton = findViewById(R.id.shutdownButton)

        // Progress Button
        mProgressButton = findViewById(R.id.indeterminateBar)

    }

    fun onShutDownButtonClicked(view: View) {
        showDialogUIElement(ALERT_ID)
    }

    // Show desired Dialog
    private fun showDialogUIElement(dialogID: Int) {

        when (dialogID) {

            // Show AlertDialog
            ALERT_ID -> {

                // Create a new AlertDialogFragment
                mDialog = AlertDialogFragment.newInstance()

                // Show AlertDialogFragment
                mDialog.show(supportFragmentManager, ALERT_TAG)
            }

            // Show ProgressDialog
            PROGRESS_ID ->
                //Make ProgressBar visible
                mProgressButton.visibility = View.VISIBLE
        }
    }

    // Abort or complete ShutDown based on value of shouldContinue
    internal fun continueShutdown(shouldContinue: Boolean) {
        if (shouldContinue) {

            // Prevent further interaction with the ShutDown Button
            mShutdownButton.isEnabled = false


            // Show ProgressDialog as shutdown process begins
            showDialogUIElement(PROGRESS_ID)

            // Finish the ShutDown process
            finishShutdown()

        } else {

            // Abort ShutDown and dismiss dialog
            mDialog.dismiss()
        }
    }

    private fun finishShutdown() {

        GlobalScope.launch(context = Dispatchers.Main) {
            delay(5000)
            finish()
        }
    }
}